core: move option results to the World class instead of MultiWorld

alwaysintreble opened the pull request:
## What is this fixing or adding?
option access go brrr

## How was this tested?
converted ror2 to using this system and generated/played seeds with it

## If this makes graphical changes, please attach screenshots.

alwaysintreble commented:
oh i forgot to make special case handling for lttp it appears. Don't really want to do that so unsure if I'll end up doing it to get this in earlier or if this gets put off until after more lttp refactoring 😓

black-sliver commented on worlds/AutoWorld.py:
```diff
@@ -121,6 +121,7 @@ class World(metaclass=AutoWorldRegister):
     A Game should have its own subclass of World in which it defines the required data structures."""
 
     option_definitions: Dict[str, Option[Any]] = {}  # link your Options mapping
+    options: Dict[str, Option[Any]] = {}  # option names to resulting option object
```
this is a static member, i.e. it will have each player of the class use the same options. you should assign a blank dict in `__init__`

If you want it to be a typed dict, this will have to create/overwrite the init per world-class in the metaclass magic, but i'd still fix this behaviour so it's easier to test.

black-sliver commented on BaseClasses.py:
```diff
@@ -204,10 +204,11 @@ def set_options(self, args):
+            self.worlds[player] = world_type(self, player)
             for option_key in world_type.option_definitions:
                 setattr(self, option_key, getattr(args, option_key, {}))
+                self.worlds[player].options[option_key] = getattr(args, option_key)[player]
```
should probably do the getattr once into a temp and then assign to both

black-sliver reviewed (changes requested):
See below. Also please update docs/world api.md

black-sliver commented on BaseClasses.py:
```diff
@@ -196,18 +196,28 @@ def set_seed(self, seed: Optional[int] = None, secure: bool = False, name: Optio
+                    self.worlds[player].options[option_key] = option_values[player]
+
             for option_key in world_type.option_definitions:
-                setattr(self, option_key, getattr(args, option_key, {}))
```
```python
for option_key in itertools.chain(Options.common_options, Options.per_game_common_options, 
                                  world_type.option_definitions):
```
can be used to get rid of ~~duplicate~~ tripled code

black-sliver commented:
I may have found another issue: `test/general/__init__.py` calls set_default_common_options - so we may have to modify that as well to provide self.options for that case.

alwaysintreble commented:
> I may have found another issue: `test/general/__init__.py` calls set_default_common_options - so we may have to modify that as well to provide self.options for that case.

Correct me if I'm wrong here, but it looks like the changes made to `set_options` covers this as that gets called first? And then we should just be able to remove `set_default_common_options` with the removal of `setattr`?

black-sliver commented on test/general/TestFill.py:
```diff
@@ -20,8 +25,12 @@ def generate_multi_world(players: int = 1) -> MultiWorld:
                         "Menu Region Hint", player_id, multi_world)
         multi_world.regions.append(region)
 
+        for option_key in itertools.chain(Options.common_options, Options.per_game_common_options):
+            option_value = getattr(args, option_key, {})
+            setattr(multi_world, option_key, option_value)
+            multi_world.worlds[player_id].options[option_key] = option_value[player_id]
+
```
Shouldn't `world.set_options(args)` correctly set up the world with args coming from `default_common_options`?

alwaysintreble replied:
The worlds get instantiated in `set_options` but `TestFill` builds the worlds so I can't just call `set_options` in it. `TestFill` is a bit janky tbh

black-sliver commented on BaseClasses.py:
```diff
@@ -262,12 +262,13 @@ def set_item_links(self):
-            setattr(self, option_key, {player_id: option(option.default) for player_id in self.player_ids})
+    @property
+    def default_common_options(self) -> Namespace:
         self.state = CollectionState(self)
```
i think this line has to move somewhere else. cleanest would probably be to move it outside of the multiworld class (into the tests), since that's what generate/main currently does here: https://github.com/ArchipelagoMW/Archipelago/blob/cded105e6c881db3b76b39279e4fb1a6829fb123/Main.py#L76-L79

alwaysintreble replied:
Can we not just create CollectionState in the init? I don't know that it's possible, but it'd certainly solve this issue for both cases.

beauxq commented on worlds/AutoWorld.py:
```diff
@@ -121,6 +121,7 @@ class World(metaclass=AutoWorldRegister):
     A Game should have its own subclass of World in which it defines the required data structures."""
 
     option_definitions: Dict[str, Option[Any]] = {}  # link your Options mapping
+    options: Dict[str, Option[Any]]  # option names to resulting option object
```
This gets auto-populated right?

A few lines below this, there's an example of a property with a comment "gets automatically populated".
I think that comment is important to make it clear to game implementers that they don't need to put anything in this.

black-sliver commented:
I have not checked this, but I think if a world writes to self.world.option_name[player], that changes spoiler output?
writes to self.options[name] are not synchronized back

alwaysintreble commented:
> I have not checked this, but I think if a world writes to self.world.option_name[player], that changes spoiler output? writes to self.options[name] are not synchronized back

This is correct actually. To support back compatibility until we rip out attr, the world options could be updated again right before output using the attributes, but this won't work with worlds that actually move over to this new system, so the only way to do this cleanly is to completely rip out the old system imo and I was kind of hoping not to do that quite yet as that is quite a bit more work and can't be easily refactored.

beauxq commented on docs/world api.md:
```diff
@@ -85,7 +85,7 @@ inside a World object.
 
 Players provide customized settings for their World in the form of yamls.
-Those are accessible through `self.world.<option_name>[self.player]`. A dict
+Those are accessible through `self.world.options["<option_name>"]`. A dict
```
The documentation being replaced specifies the player. Shouldn't the replacement also specify the player?

beauxq commented on docs/world api.md:
````diff
@@ -455,7 +456,7 @@ In addition, the following methods can be implemented and attributes can be set
 ```python
 def generate_early(self) -> None:
     # read player settings to world instance
-    self.final_boss_hp = self.world.final_boss_hp[self.player].value
+    self.final_boss_hp = self.world.options["final_boss_hp"].value
````
similar to previous comment: How does it know which player's `final_boss_hp` to get the value from?

alwaysintreble replied:
These dictionaries exist per `World` so this dictionary is all of the options that your current world's player chose.

beauxq replied:
What is the `self` that has a `world` on it? This doesn't include the PR that changes `world` to `multiworld`, right?

I think that's a `MultiWorld` that the documentation is referencing.

alwaysintreble replied:
it's a typo :)

Berserker66 commented:
> > I have not checked this, but I think if a world writes to self.world.option_name[player], that changes spoiler output? writes to self.options[name] are not synchronized back
> 
> This is correct actually. To support back compatibility until we rip out attr, the world options could be updated again right before output using the attributes, but this won't work with worlds that actually move over to this new system, so the only way to do this cleanly is to completely rip out the old system imo and I was kind of hoping not to do that quite yet as that is quite a bit more work and can't be easily refactored.

There's quite a few worlds that write back, the classic is having min higher than a max and then setting one to the other, so a solution for that should be found.

alwaysintreble commented:
As per @el-u 's comment I did a quick sweep to check for games that were overriding the option by creating a new option class and changed those to just change the value instead, so all the not updated worlds should still write to spoiler correctly, though I haven't tested it thoroughly yet. Updated RoR2 to use the new system again, and tested that `as_dict` was working with it.

kindasneaki commented on worlds/ror2/Options.py:
```diff
@@ -271,39 +271,40 @@ class ItemWeights(Choice):
+    total_revivals: TotalRevivals
+    start_with_revive: StartWithRevive
+    final_stage_death: FinalStageDeath
+    being_with_loop: BeginWithLoop
```
this should be begin_with_loop

alwaysintreble commented on worlds/ror2/__init__.py:
```diff
@@ -225,20 +225,20 @@ def create_item(self, name: str) -> Item:
 
 def create_events(world: MultiWorld, player: int) -> None:
-    total_locations = world.total_locations[player].value
+    total_locations = world.worlds[player].options.total_locations.value
```
I agree but it would require me to rewrite part of the World to do that and it isn't my world.

beauxq commented on docs/world api.md:
```diff
@@ -374,15 +374,10 @@ class MyGameLocation(Location):  # or from Locations import MyGameLocation
-    # cache any texts, then increase by 1 for each release that makes changes.
-    data_version = 0
+    game = "My Game"  # name of the game/world
+    options_dataclass = MyGameOptions # options the player can set
```
```suggestion
    options_dataclass = MyGameOptions  # options the player can set
```
I think pep 8 calls for 2 spaces before a comment.

ThePhar commented:
![image](https://github.com/ArchipelagoMW/Archipelago/assets/11338376/8739a031-8907-4548-9125-22fb4638cd91)

agilbert1412 reviewed (approved):
I looked through the part of the PR that touches stardew, and quite frankly I don't understand neither the original code nor your changes.

If the tests still run, pass, and cover the same things, I don't have a problem with it. I have to trust that you know what you're doing. I cannot offer an opinion of value here, but the changes seem very minimal.

kindasneaki reviewed (approved):
Looked through the RoR2 stuff and everything looks good to me

ScipioWright reviewed (approved):
Checked it out with assistance from Treble and eL. First off, it's backwards compatible, so it shouldn't create issues with existing worlds. New system is cleaner looking, and gets rid of those annoying underlines for options. Will absolutely swap over Noita to the new system when it gets merged, especially since it's already partially done being moved over to it in a local branch.

BootsinSoots commented:
So, the way I understand what we are doing here, is that we are moving options into their own class pulled from the options file, such that they are no longer stored on the multiworld object. I am curious as to why we are doing this - not because I see an issue necessarily but because I assume there is a purpose in doing so. It also seems that switching to this should be as simple as replacing `multiworld` with `options` and remove the player reference. Does this mean that, if I use my options when determining access_rules, generation knows which player world has which setting (in the case of multiple worlds of the same game)?

beauxq commented:
> I am curious as to why we are doing this

One of the biggest reasons I want this is for the type safety.
Type checkers can't have any idea what option objects are on the `MultiWorld` class, so you have to hope that your tests didn't miss a typo in the option name or something like that.
With this change, type checkers can make sure that everything you're getting out of your options is being used in type-appropriate ways.

beauxq commented on docs/options api.md:
```diff
@@ -28,47 +28,54 @@ Choice, and defining `alias_true = option_full`.
 Options are parsed by `Generate.py` before the worlds are created, and then the option classes are created shortly after
 world instantiation. These are created as attributes on the MultiWorld and can be accessed with
-`self.multiworld.my_option_name[self.player]`. This is the option class, which supports direct comparison methods to
+`self.o.my_option_name`. This is the option class, which supports direct comparison methods to
```
```suggestion
`self.options.my_option_name`. This is an instance of the option class, which supports direct comparison methods to
```
note and double-check the 2 changes here:

`self.o` or `self.options`?

Normally I wouldn't be so pedantic between "instance" and "class", but in this case, we do a lot with the class objects that aren't instances of the class, so it's important to note the difference.

beauxq commented:
@alwaysintreble https://github.com/alwaysintreble/Archipelago/pull/17

beauxq reviewed (approved):
I have spent a bunch of time looking at the core changes.
Issues I've seen have been addressed.

BootsinSoots commented:
>  Does this mean that, if I use my options when determining access_rules, generation knows which player world has which setting (in the case of multiple worlds of the same game)?

Never really got a response on this, just wanted to make sure it IS, in fact, this simple when using lambdas in `access_rule`

black-sliver reviewed (approved):
Last time I checked, code lgtm. We have multiple approvals. I have not tested a ton, ~~but that's what main is for~~ but pushing it further back does not sound like a solution.

t3hf1gm3nt reviewed (approved):
fully admit i still dont fully understand what all is going on with this, but have tested that TLOZ gens fine under this change (genned a few games and checked spoilers that options were working correctly) so LGTM

axe-y reviewed (approved):
Unsure of what is happening in /Options but for everything else good.

kindasneaki reviewed (changes requested):
`from .Options import DLCQuest_options, DLCQuestOptions, fetch_options`
both `DLCQuest_options` and `fetch_options` dont exist anymore and shouldn't be imported

beauxq commented on Options.py:
```diff
@@ -955,6 +955,14 @@ def as_dict(self, *option_names: str, casing: str = "snake") -> typing.Dict[str,
                 raise ValueError(f"{option_name} not found in {tuple(type(self).type_hints)}")
         return option_results
 
+    def get_value(self, option_name: str) -> typing.Any:
```
I'm also not feeling good about this. It could lead people to do things in ways that aren't type safe.

If this is needed, I would like to see in the doc string a recommendation against using it.

agilbert1412 replied:
What would you propose as a way to access a single option by name?

alwaysintreble replied:
In the actual context of when this is used, I think it's fine. But yeah, this is just ripe for people copy pasting it, so I can agree with a doc string recommending against that, at minimum.

beauxq replied:
> What would you propose as a way to access a single option by name?

PR incoming

beauxq replied:
> What would you propose as a way to access a single option by name?

https://github.com/alwaysintreble/Archipelago/pull/21

black-sliver commented on docs/world api.md:
```diff
@@ -702,6 +712,26 @@ def generate_output(self, output_directory: str):
+### Slot Data
+
+If the game client needs to know information about the generated seed, a preferred method of transferring the data
+is through the slot data. This can be filled in the `fill_slot_data` method on your world. This can be used to fill any
+data you need to know client side, and should return a `Dict[Any]`, but should be limited to only data that is
+absolutely necessary to not waste data. Slot data is sent to your client once it has successfully
+[connected.](network%20protocol.md#connected) If you need to know information about locations in your world, instead
```
```suggestion
is through the slot data. This can be filled from the `fill_slot_data` method of your world by returing a `Dict[Any]`,
but should be limited to data that is absolutely necessary to not waste resources. Slot data is sent to your client once
it has successfully [connected](network%20protocol.md#connected).
If you need to know information about locations in your world, instead
```

alwaysintreble replied:
returing

alwaysintreble replied:
it's technically `Dict[Any, Any]` we just usually use `Dict[str, Any]`

black-sliver replied:
fixed the suggestion. Can't be [Any, Any] because of json

beauxq replied:
The slot data needs to be JSON serializable.
The typing of JSON data has been discussed at length in other places https://github.com/python/typing/issues/182
Because of its complexity, I think it would be better to not mention the type and just mention that it needs to be JSON serializable.
